src/jit/codegenarm64.cpp

```diff
@@ -6476,287 +6483,388 @@ void CodeGen::genPutArgStk(GenTreePtr treeNode)
 
-        // We will need to record the GC type used by each of the load instructions
-        //  so that we use the same type in each of the store instructions
-        var_types type0 = TYP_UNKNOWN;
-        var_types type1 = TYP_UNKNOWN;
+            GenTreeLclVarCommon*  varNode  = nullptr;
+            GenTreePtr            addrNode = nullptr;
 
-        if (data->OperGet() == GT_OBJ)
-        {
-            GenTree* objNode  = data;
-            GenTree* addrNode = objNode->gtOp.gtOp1;
-
-            if (addrNode->OperGet() == GT_LCL_VAR_ADDR)
+            if (source->OperGet() == GT_LCL_VAR)
             {
-                // We have a GT_OBJ(GT_LCL_VAR_ADDR)
-                //
-                // We will treat this case the same as a GT_LCL_VAR node 
-                // so update 'data' to point this GT_LCL_VAR_ADDR node
-                // and continue to the codegen for the LCL_VAR node below
+                varNode = source->AsLclVarCommon();
+            }
+            else // we must have a GT_OBJ
+            {
+                assert(source->OperGet() == GT_OBJ);
+
+                addrNode = source->gtOp.gtOp1;
+
+                // addrNode can either be a GT_LCL_VAR_ADDR or an address expression
                 //
-                data = addrNode;
+                if (addrNode->OperGet() == GT_LCL_VAR_ADDR)
+                {
+                    // We have a GT_OBJ(GT_LCL_VAR_ADDR)
+                    //
+                    // We will treat this case the same as above 
+                    // (i.e if we just had this GT_LCL_VAR directly as the source)
+                    // so update 'source' to point this GT_LCL_VAR_ADDR node
+                    // and continue to the codegen for the LCL_VAR node below
+                    //
+                    varNode  = addrNode->AsLclVarCommon();
+                    addrNode = nullptr;
+                }
             }
-            else  // We have a GT_OBJ with an address expression
+
+            // Either varNode or addrNOde must have been setup above,
+            // the xor ensures that only one of the two is setup, not both
+            assert((varNode != nullptr) ^ (addrNode != nullptr));
+
+            BYTE     gcPtrs[MAX_MULTIREG_ARGS] = {TYPE_GC_NONE, TYPE_GC_NONE, TYPE_GC_NONE, TYPE_GC_NONE };
+
+            BYTE*    structGcLayout = &gcPtrs[0];  // The GC layout for the struct
+            unsigned gcPtrCount;                   // The count of GC pointers in the struct
+            int      structSize;
+            bool     isHfa;
+
+            // Setup the structSize, isHFa, and gcPtrCount
+            if (varNode != nullptr)
             {
+                varNumInp = varNode->gtLclNum;
+                assert(varNumInp < compiler->lvaCount);
+                LclVarDsc* varDsc = &compiler->lvaTable[varNumInp];
+
+                assert(varDsc->lvType = TYP_STRUCT);
+                assert(varDsc->lvOnFrame);       // This struct also must live in the stack frame
```

Comment:
Don't we also need to assert that varDsc is not in a register.  Think of SIMD types (or intrinsics) enabled.